system/security.php

```diff
@@ -0,0 +1,422 @@
+             * Let's grab the necessary number of random bytes
+             */
+            $randomByteString = random_bytes($bytes);
+            if ($randomByteString === false) {
+                throw new Exception('Random number generator failure');
+            }
+
+            /**
+             * Let's turn $randomByteString into an integer
+             *
+             * This uses bitwise operators (<< and |) to build an integer
+             * out of the values extracted from ord()
+             *
+             * Example: [9F] | [6D] | [32] | [0C] =>
+             *   159 + 27904 + 3276800 + 201326592 =>
+             *   204631455
+             */
+            $val = 0;
+            for ($i = 0; $i < $bytes; ++$i) {
+                $val |= ord($randomByteString[$i]) << ($i * 8);
+            }
+
+            /**
+             * Apply mask
+             */
+            $val &= $mask;
+            $val += $valueShift;
+
+            ++$attempts;
+            /**
+             * If $val overflows to a floating point number,
+             * ... or is larger than $max,
+             * ... or smaller than $int,
+             * then try again.
+             */
+        } while (!is_int($val) || $val > $max || $val < $min);
+        return (int) $val;
+    }
+
+    /**
+     * Generate a random string
+     *
+     * @param int $length How many characters long should it be?
+     * @param string $alphabet Which characters should we allow?
+     *
+     * @return string
+     */
+    public static function randomString($length = 16, $alphabet = 'abcdefghijklmnopqrstuvwxyz')
+    {
+        if ($length < 1) {
+            throw new InvalidArgumentException('Length must be a positive integer');
+        }
+        $str = '';
+        $alphmax = self::ourStrlen($alphabet) - 1;
+        if ($alphamax < 1) {
+            throw new InvalidArgumentException('Invalid alphabet');
+        }
+        for ($i = 0; $i < $length; ++$i) {
+            $str .= $alphabet[self::randomInt(0, $alphmax)];
+        }
+        return $str;
+    }
+
+    /**
+     * Compare two strings without timing side-channels
+     *
+     * @staticvar boolean $native Use native hash_equals()? (PHP 5.6+ only)
+     *
+     * @param string $expected string (raw binary)
+     * @param string $given string (raw binary)
+     * @return boolean
+     */
+    public static function hashEquals($expected, $given)
+    {
+        static $native = null;
+        if ($native === null) {
+            $native = function_exists('hash_equals');
+        }
+        if ($native) {
+            return hash_equals($expected, $given);
+        }
+
+        if (self::safeStrlen($expected) !== self::safeStrlen($given)) {
+            return false;
+        }
+        $blind = self::randomBytes(32);
+        $message_compare = hash_hmac('sha256', $given, $blind);
+        $correct_compare = hash_hmac('sha256', $expected, $blind);
+        return $correct_compare === $message_compare;
+    }
```

Comment:
Good question. This is a timing attack mitigation strategy called "Double HMAC". By using a random nonce (`$blind` above), you eliminate any useful timing information from leaking out.